File: 0794-swim-in-rising-water/0794-swim-in-rising-water.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var swimInWater = function(grid) {
    const n = grid.length;
    const heap = [[grid[0][0], 0, 0]];
    const visited = new Set();
    const directions = [[-1, 0], [1, 0], [0, -1], [0, 1]];
    
    while (heap.length > 0) {
        heap.sort((a, b) => a[0] - b[0]);
        const [maxElev, row, col] = heap.shift();
        
        if (row === n - 1 && col === n - 1) {
            return maxElev;
        }
        
        const key = `${row},${col}`;
        if (visited.has(key)) continue;
        
        visited.add(key);
        
        for (const [dr, dc] of directions) {
            const newRow = row + dr;
            const newCol = col + dc;
            const newKey = `${newRow},${newCol}`;
            
            if (newRow >= 0 && newRow < n && newCol >= 0 && newCol < n && 
                !visited.has(newKey)) {
                const newMaxElev = Math.max(maxElev, grid[newRow][newCol]);
                heap.push([newMaxElev, newRow, newCol]);
            }
        }
    }
    
    return -1;
};

File: 0794-swim-in-rising-water/0794-swim-in-rising-water.cpp
class Solution {
public:
    int swimInWater(vector<vector<int>>& grid) {
        int n = grid.size();
        priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>>  pq;
        pq.push({grid[0][0], 0, 0});

        vector<vector<bool>> visited(n, vector<bool>(n, false));
        vector<pair<int, int>> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        while (!pq.empty()) {
            auto curr = pq.top();
            pq.pop();

            int maxElev = curr[0];
            int row = curr[1];
            int col = curr[2];

            if (row == n - 1 && col == n - 1) {
                return maxElev;
            }
            if (visited[row][col])
                continue;
            visited[row][col] = true;

            for (auto& [dr, dc] : directions) {
                int newRow = row + dr;
                int newCol = col + dc;

                if (newRow >= 0 && newRow < n && newCol >= 0 && newCol < n &&
                    !visited[newRow][newCol]) {
                    int newMaxElev = max(maxElev, grid[newRow][newCol]);
                    pq.push({newMaxElev, newRow, newCol});
                }
            }
        }

        return -1;
    }
};